docs/advanced/cast/custom.rst

```diff
@@ -64,7 +64,7 @@ type is explicitly allowed.
                 value.long_value = PyLong_AsLong(tmp);
                 Py_DECREF(tmp);
                 /* Ensure return code was OK (to avoid out-of-range errors etc) */
-                return !(value.long_value == -1 && !PyErr_Occurred());
+                return !(value.long_value == -1 && PyErr_Occurred());
```

Comment:
The error needs to be cleared, too.

A better version of custom.rst file is buried in here:

https://github.com/pybind/pybind11/pull/3862/files#diff-56bf050c9c6d51442eb6561df1438ea1ce5d05d5f3b339f36a72a779155131b0

The ADL approach under #3862 didn't work out, so the parts referring to that need to be removed.

Looking at it with a fresh eye, I think there is still a bug in the `cast()` implementation: missing error handling again.

Let me see if I can rescue custom.rst and tests/test_docs_advanced_cast_custom.cpp,py from the old PR. Maybe this weekend.

Reply:
Or please let me know if you want to help with the rescuing.

Reply:
Well, I'd be happy to give a shot at fully sanitizing the custom type caster example. I can add the other missing error checks to this PR, also including the associated modifications of the test.

Reply:
Awesome, please tag me here when this PR is ready for re-review. It'll be great if you could adopt the doc changes from 3862 as well (excluding the ADL-related bits).